feat: gameloop addition

Sloox opened the pull request:
Fixes #1245 

adds the gameloop into android (androidtestloop)


## Checklist

- [x] Documented
- [x] Unit tested

jan-gogo commented on test_runner/src/main/kotlin/ftl/gc/GcAndroidTestMatrix.kt:
```diff
@@ -56,7 +56,7 @@ object GcAndroidTestMatrix {
             .setDirectoriesToPull(args.directoriesToPull)
             .setAdditionalApks(additionalApkGcsPaths.mapGcsPathsToApks())
-            .setFilesToPush(otherFiles.mapToDeviceFiles())
+            .setFilesToPush(otherFiles.mapToDeviceFiles() + args.obbFiles.mapToDeviceFiles())
```
`obbFiles` need to be uploaded to bucket like other files https://github.com/Flank/flank/blob/b9596bc83027a5bc94bb5d1dca9a8adb81dcea5a/test_runner/src/main/kotlin/ftl/run/platform/android/UploadOtherFiles.kt#L10

Sloox replied:
will investigate and add whats needed.

Sloox replied:
resolved

jan-gogo commented on test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:
```diff
@@ -91,24 +91,27 @@ AndroidArgs
             additionalAppTestApks.isEmpty()
 
+val AndroidArgs.isGameLoop
+    get() = if (type == null) false else (type == Type.GAMELOOP)
```
I don't like `type` flag, it's redundant in 2/3 and only really, required for gameloop. Is there any easy way to identify gameloop without this flag? If not. skip this comment.

Sloox replied:
No there isnt an easy way. The type flag appears to be what probably should have been used to identify the robo, instrumental and gameloop in the first place.

jan-gogo commented on test_runner/src/main/kotlin/ftl/gc/android/Utils.kt:
```diff
@@ -17,3 +18,10 @@ internal fun Map<String, String>.mapToDeviceFiles(): List<DeviceFile> =
         )
     }
+
+internal fun List<String>.mapToDeviceFiles(): List<DeviceFile> =
```
`mapToDeviceFiles` -> `mapToDeviceObbFiles`

jan-gogo commented on test_runner/src/main/kotlin/ftl/run/platform/android/AndroidTestConfig.kt:
```diff
@@ -22,4 +22,12 @@ sealed class AndroidTestConfig {
+
+    data class GameLoop(
+        val appApkGcsPath: String,
+        val testApkGcsPath: String,
```
Not used anywhere

pawelpasterz reviewed (changes requested):
I've noticed some differences in logic compared with `gcloud`

1. ```
      app: [path]/app-debug.apk
      test: [path]/app-multiple-success-debug-androidTest.apk
      type: game-loop
    ```
    runs normal instrumentation test where gcloud throw error
    ```
    ERROR: (gcloud.alpha.firebase.test.android.run) Invalid value for [test]: may not be used with test type [game-loop]
    ```

1. ```
    app: [path]/app-debug.apk
    type: game-loop
    ```
    runs sanity robo where `gcloud` runs `game-loop`

1. ```
    app: [path]/app-debug.apk
    obb-files:
      - [path]/test.obb <== the same as in PR
    type: game-loop
    ```
   
    results in a nasty exception
    ```
    RunTests
      Uploading app-debug.apk .
      java.io.IOException: Request failed
	    at ftl.http.ExecuteWithRetryKt$executeWithRetry$$inlined$withRetry$1.invokeSuspend(ExecuteWithRetry.kt:40)
	    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	    at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:176)
	    at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:111)
	    at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:308)
	    at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:318)
	    at kotlinx.coroutines.CancellableContinuationImpl.resumeUndispatched(CancellableContinuationImpl.kt:400)
	    at kotlinx.coroutines.EventLoopImplBase$DelayedResumeTask.run(EventLoop.common.kt:489)
	    at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
 	    at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:84)
	    at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	    at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	    at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	    at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	    at ftl.http.ExecuteWithRetryKt.executeWithRetry(ExecuteWithRetry.kt:45)
	    at ftl.run.platform.RunAndroidTestsKt$executeAndroidTestMatrix$2$invokeSuspend$$inlined$map$lambda$1.invokeSuspend(RunAndroidTests.kt:103)
	    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
	    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
	    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
    Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
    POST https://testing.googleapis.com/v1/projects/pawel-p/testMatrices
    {
      "code" : 400,
      "errors" : [ {
        "domain" : "global",
        "message" : "Required field should not be empty: google.devtools.testing.v1.ObbFile.obb_file_name",
        "reason" : "badRequest"
      } ],
      "message" : "Required field should not be empty: google.devtools.testing.v1.ObbFile.obb_file_name",
      "status" : "INVALID_ARGUMENT"
    }
	    at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:149)
	    at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:112)
	    at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:39)
	    at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1.interceptResponse(AbstractGoogleClientRequest.java:443)
	    at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1108)
	    at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:541)
	    at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:474)
	    at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:591)
	    at ftl.http.ExecuteWithRetryKt$executeWithRetry$$inlined$withRetry$1.invokeSuspend(ExecuteWithRetry.kt:45)
	    ... 21 more
   ```
    and `gcloud`'s
    ```
    ERROR: (gcloud.alpha.firebase.test.android.run) Invalid value for [obb-files]: 
    [path/test.obb] is not a valid OBB file name, which must have the 
    format: (main|patch).<versionCode>.<package.name>.obb
    ```

1. ```
    app: [path]/app-debug.apk
    obb-files:
      -  [path]/main.0300110.com.example.android.obb
    type: game-loop
    ```
    here is interesting because flank throws an exception the same as above where gcloud results:
   ```
    ERROR: (gcloud.alpha.firebase.test.android.run)
Matrix [matrix-2p3op1ukymruk] failed during validation: The app does not have a correctly formatted game-loop intent filter.
    ```


I might be wrong and some of those errors are due to my incorrect configuration. Please verify

Sloox commented on test_runner/src/main/kotlin/ftl/run/model/AndroidTestContext.kt:
```diff
@@ -19,6 +19,13 @@ data class RoboTestContext(
 
+data class GameLoopContext(
+    val app: FileReference,
+    val test: FileReference?,
```
removed

piotradamczyk5 commented on test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:
```diff
@@ -91,24 +91,27 @@ AndroidArgs
-            (appApk.isNotNull() || additionalAppTestApks.all { (app, _) -> app.isNotNull() })
+    get() = if (type != null) (type == Type.INSTRUMENTATION) else
+        (appApk.isNotNull() && testApk.isNotNull() || additionalAppTestApks.isNotEmpty() &&
+            (appApk.isNotNull() || additionalAppTestApks.all { (app, _) -> app.isNotNull() }))
```
IMHO  with this formatting this is not so readable I would use something like
```kotlin
get() = if (type != null) (type == Type.INSTRUMENTATION) else validateApks()

fun AndroidArgs.validateApks() = (appApk.isNotNull() && testApk.isNotNull() || 
    additionalAppTestApks.isNotEmpty() &&
    (appApk.isNotNull() || additionalAppTestApks.all { (app, _) -> app.isNotNull() }))
```

Sloox replied:
👍

piotradamczyk5 commented on test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:
```diff
@@ -91,24 +91,27 @@ AndroidArgs
-    get() = appApk.isNotNull() &&
-            (roboDirectives.isNotEmpty() || roboScript.isNotNull())
+    get() = if (type != null) (type == Type.ROBO) else
+        (appApk.isNotNull() && (roboDirectives.isNotEmpty() || roboScript.isNotNull()))
```
this is better, but also could be wrap in `validateRobo` function

Sloox replied:
👍

piotradamczyk5 commented on test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:
```diff
@@ -91,24 +91,27 @@ AndroidArgs
 val AndroidArgs.isRoboTest
-    get() = appApk.isNotNull() &&
-            (roboDirectives.isNotEmpty() || roboScript.isNotNull())
+    get() = if (type != null) (type == Type.ROBO) else
```
```suggestion
    get() = if (type != null) type == Type.ROBO else
```

`()` not needed

Sloox replied:
👍

piotradamczyk5 commented on test_runner/src/main/kotlin/ftl/args/ValidateAndroidArgs.kt:
```diff
@@ -40,12 +40,12 @@ private fun AndroidArgs.assertLabelContent() {
+    if (obbFiles.isNotEmpty()) {
+        obbFiles.forEach {
             ArgsHelper.assertFileExists(it, " (obb file)")
         }
```
You have triple check of `obbFiles.isNotEmpty`.
How about?
```kotlin
when {
    obbFiles.isEmpty() -> Unit
    (type == null || type != Type.GAMELOOP) -> throw FlankConfigurationError("OBB files defined but Type is not Game-loop. ($type)")
    obbFiles.size > MAX_OBB_FILES -> throw FlankConfigurationError("Up to two OBB files are supported. Currently ${obbFiles.size} OBB files have been supplied.")
    else -> obbFiles.forEach { ArgsHelper.assertFileExists(it, " (obb file)")  }
}
...
private const val MAX_OBB_FILES = 2
```

Sloox replied:
👍

piotradamczyk5 commented on test_runner/src/main/kotlin/ftl/gc/android/SetupAndroidTest.kt:
```diff
@@ -9,5 +9,7 @@ internal fun TestSpecification.setupAndroidTest(config: AndroidTestConfig) = app
         is AndroidTestConfig.Robo ->
             androidRoboTest = createAndroidRoboTest(config)
+        is AndroidTestConfig.GameLoop ->
+            androidTestLoop = createGameLoopTest(config)
```
this is very short and I deb that it could be a single line :)

Sloox replied:
👍

piotradamczyk5 commented on test_runner/src/main/kotlin/ftl/gc/android/Utils.kt:
```diff
@@ -17,3 +18,10 @@ internal fun Map<String, String>.mapToDeviceFiles(): List<DeviceFile> =
     }
+
+internal fun Map<String, String>.mapToDeviceObbFiles(): List<DeviceFile> =
+    map { (_: String, gcsFilePath: String) ->
```
```suggestion
    map { (_, gcsFilePath) ->
```

Sloox replied:
👍

piotradamczyk5 reviewed (changes requested):
please address comments and update `docs/index.md` and `flank.yml`

Sloox commented:
> please address comments and update `docs/index.md` and `flank.yml`

its already done apart of other tickets. This topic merely forwards and enables the previous tickets work.

Sloox commented:
> I've noticed some differences in logic compared with `gcloud`
> 
> 1. ```
>       app: [path]/app-debug.apk
>       test: [path]/app-multiple-success-debug-androidTest.apk
>       type: game-loop
>    ```
>    
>    
>    runs normal instrumentation test where gcloud throw error
>    ```
>    ERROR: (gcloud.alpha.firebase.test.android.run) Invalid value for [test]: may not be used with test type [game-loop]
>    ```
> 2. ```
>     app: [path]/app-debug.apk
>     type: game-loop
>    ```
>    
>    
>    runs sanity robo where `gcloud` runs `game-loop`
> 3. ```
>     app: [path]/app-debug.apk
>     obb-files:
>       - [path]/test.obb <== the same as in PR
>     type: game-loop
>    ```
>    
>    
>    results in a nasty exception
>    ```
>    RunTests
>      Uploading app-debug.apk .
>      java.io.IOException: Request failed
>        at ftl.http.ExecuteWithRetryKt$executeWithRetry$$inlined$withRetry$1.invokeSuspend(ExecuteWithRetry.kt:40)
>        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
>        at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:176)
>        at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:111)
>        at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:308)
>        at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:318)
>        at kotlinx.coroutines.CancellableContinuationImpl.resumeUndispatched(CancellableContinuationImpl.kt:400)
>        at kotlinx.coroutines.EventLoopImplBase$DelayedResumeTask.run(EventLoop.common.kt:489)
>        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
>        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:84)
>        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
>        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
>        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
>        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
>        at ftl.http.ExecuteWithRetryKt.executeWithRetry(ExecuteWithRetry.kt:45)
>        at ftl.run.platform.RunAndroidTestsKt$executeAndroidTestMatrix$2$invokeSuspend$$inlined$map$lambda$1.invokeSuspend(RunAndroidTests.kt:103)
>        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
>        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
>        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
>        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
>        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
>        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
>    Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
>    POST https://testing.googleapis.com/v1/projects/pawel-p/testMatrices
>    {
>      "code" : 400,
>      "errors" : [ {
>        "domain" : "global",
>        "message" : "Required field should not be empty: google.devtools.testing.v1.ObbFile.obb_file_name",
>        "reason" : "badRequest"
>      } ],
>      "message" : "Required field should not be empty: google.devtools.testing.v1.ObbFile.obb_file_name",
>      "status" : "INVALID_ARGUMENT"
>    }
>        at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:149)
>        at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:112)
>        at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:39)
>        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1.interceptResponse(AbstractGoogleClientRequest.java:443)
>        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1108)
>        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:541)
>        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:474)
>        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:591)
>        at ftl.http.ExecuteWithRetryKt$executeWithRetry$$inlined$withRetry$1.invokeSuspend(ExecuteWithRetry.kt:45)
>        ... 21 more
>    ```
>    
>    
>    and `gcloud`'s
>    ```
>    ERROR: (gcloud.alpha.firebase.test.android.run) Invalid value for [obb-files]: 
>    [path/test.obb] is not a valid OBB file name, which must have the 
>    format: (main|patch).<versionCode>.<package.name>.obb
>    ```
> 4. ```
>     app: [path]/app-debug.apk
>     obb-files:
>       -  [path]/main.0300110.com.example.android.obb
>     type: game-loop
>    ```
>    
>    
>    here is interesting because flank throws an exception the same as above where gcloud results:
>    ```
>     ERROR: (gcloud.alpha.firebase.test.android.run)
>     Matrix [matrix-2p3op1ukymruk] failed during validation: The app does not have a correctly formatted game-loop intent 
>     filter.
>    ```
> 
> I might be wrong and some of those errors are due to my incorrect configuration. Please verify
To address each question
1. test is removed as per @jan-gogo change request.
2. unless I have tested incorrectly i am not getting the same behavior as you mention it is running a AndroidLoop. Can you verify
3. Fixed and added a required addition where the name of the obbfile is required. This is due to a mismatch of documention between ftl and the actual error message. Documentation states "can be null" but it is required.
4. No longer applicable.
Take note that the obb-names addition needs to now be added and verified and of the correct format.
If not it will fail at the ftl side.
more information about the obb files and their names: https://firebase.google.com/docs/test-lab/reference/testing/rest/v1/projects.testMatrices#obbfile
&& https://developer.android.com/google/play/expansion-files

Sloox commented:
@pawelpasterz

pawelpasterz commented:
> 2. unless I have tested incorrectly i am not getting the same behavior as you mention it is running a AndroidLoop. Can you verify

This is what I get when launched with flank
https://console.firebase.google.com/project/flank-open-source/testlab/histories/bh.da0c237aaa33732/matrices/7368025719024499574
<img width="320" alt="Screenshot 2020-10-26 at 11 47 28" src="https://user-images.githubusercontent.com/32893017/97163513-0e1e0300-1781-11eb-98af-5b08b97a1bfa.png">

and gcloud (it actually fails but you can see it's trying to run game-loop)
<img width="834" alt="Screenshot 2020-10-26 at 11 50 01" src="https://user-images.githubusercontent.com/32893017/97163770-69e88c00-1781-11eb-93f6-f4000e1e3e85.png">


copy-pasted config
```
gcloud:
  app: [path]/app-debug.apk
  type: game-loop
```

pawelpasterz commented:
I can reproduce it 10/10. Flank built from commit 
```
version: local_snapshot
revision: f91d325cbb5fe547a040d79c9925b479831cb07e
```

Sloox commented:
@pawelpasterz Added manual tests have a look at the Debug.kt. Also added a semi-compatible android application that works without crashes.  You can see the latest commit to see whats added.

pawelpasterz commented on test_runner/src/main/kotlin/ftl/gc/android/Utils.kt:
```diff
@@ -23,6 +24,16 @@ internal fun Map<String, String>.mapToDeviceFiles(): List<DeviceFile> =
 
+internal fun Map<String, String>.mapToDeviceObbFiles(obbnames: List<String>): List<DeviceFile> {
+    var index = -1 // intentional
+    return map { (_, gcsFilePath: String) ->
```
how about `mapIndexed` ?
```suggestion
    return values.mapIndexed { index, gcsFilePath -> ...
```

Sloox replied:
👍

pawelpasterz commented on test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:
```diff
@@ -90,24 +92,33 @@ AndroidArgs
+    get() = if (type != null) (type == Type.ROBO) && checkForSanityRobo else checkForSanityRobo
+
+val AndroidArgs.isGameLoop
+    get() = if (type == null) false else (type == Type.GAMELOOP)
```
I think we could simplify it:
```suggestion
    get() = type == Type.GAMELOOP
```
I think enums have `==` implemented as safe calls out of the box

Sloox replied:
I will leave this one as is.

pawelpasterz commented on test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:
```diff
@@ -90,24 +92,33 @@ AndroidArgs
+
+fun AndroidArgs.validateRobo() = (appApk.isNotNull() && (roboDirectives.isNotEmpty() || roboScript.isNotNull()))
+
+val AndroidArgs.checkForSanityRobo
```
I think `checkForSanityRobo` is not used anywhere else, let's make it private

Sloox replied:
👍
